ddtrace/tracer/sampler_test.go

```diff
@@ -179,3 +183,280 @@ func TestRateSamplerSetting(t *testing.T) {
+				assert := assert.New(t)
+				rs := newRulesSampler(v)
+
+				span := makeSpan("http.request", "test-service")
+				result := rs.apply(span)
+				assert.True(result)
+				assert.Equal(1.0, span.Metrics["_dd.rule_psr"])
+				assert.Equal(0.5, span.Metrics["_dd.limit_psr"])
+			})
+		}
+	})
+
+	t.Run("default-rate", func(t *testing.T) {
+		ruleSets := [][]SamplingRule{
+			{},
+			{ServiceRule("other-service", 0.0)},
+		}
+		for _, v := range ruleSets {
+			t.Run("", func(t *testing.T) {
+				assert := assert.New(t)
+				os.Setenv("DD_TRACE_SAMPLE_RATE", "0.8")
+				defer os.Unsetenv("DD_TRACE_SAMPLE_RATE")
+				rs := newRulesSampler(v)
+
+				span := makeSpan("http.request", "test-service")
+				result := rs.apply(span)
+				assert.True(result)
+				assert.Equal(0.8, span.Metrics["_dd.rule_psr"])
+				assert.Equal(0.5, span.Metrics["_dd.limit_psr"])
+			})
+		}
+	})
+}
+
+func BenchmarkRulesSampler(b *testing.B) {
+	b.Run("no-rules", func(b *testing.B) {
+		tracer := newTracer()
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			tracer.StartSpan("web.request", Tag(ext.SamplingPriority, ext.PriorityUserKeep)).Finish()
+		}
+	})
+
+	b.Run("unmatching-rules", func(b *testing.B) {
+		rules := []SamplingRule{
+			ServiceRule("test-service", 1.0),
+			ServiceOperationRule("postgres.db", "db.query", 1.0),
+			OperationRule("notweb.request", 1.0),
+		}
+		tracer := newTracer(WithSamplingRules(rules))
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			tracer.StartSpan("web.request", Tag(ext.SamplingPriority, ext.PriorityUserKeep)).Finish()
+		}
+	})
+
+	b.Run("matching-rules", func(b *testing.B) {
+		rules := []SamplingRule{
+			ServiceRule("test-service", 1.0),
+			ServiceOperationRule("postgres.db", "db.query", 1.0),
+			OperationRule("web.request", 1.0),
+		}
+		tracer := newTracer(WithSamplingRules(rules))
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			tracer.StartSpan("web.request", Tag(ext.SamplingPriority, ext.PriorityUserKeep)).Finish()
+		}
+	})
+}
+
+// https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go#compiler-optimisation
+var matched bool
+
+func BenchmarkRulesSamplerImpls(b *testing.B) {
+	b.Run("structs", func(b *testing.B) {
+		rules := []SamplingRule{
+			ServiceRule("test-service", 1.0),
+			ServiceOperationRule("postgres.db", "db.query", 1.0),
+			OperationRule("web.request", 1.0),
+		}
+		sampler := newRulesSampler(rules)
+		span := newSpan("web.request", "xyz-service", "", 0, 0, 0)
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			matched = sampler.apply(span)
+		}
+	})
+
+	b.Run("pointers", func(b *testing.B) {
+		rules := []SamplingRule{
+			ServiceRule("test-service", 1.0),
+			ServiceOperationRule("postgres.db", "db.query", 1.0),
+			OperationRule("web.request", 1.0),
+		}
+		sampler := newRulesSamplerP(rules)
+		span := newSpan("web.request", "xyz-service", "", 0, 0, 0)
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			matched = sampler.apply(span)
+		}
+	})
+}
+
+type rulesSamplerP struct {
```

Comment:
What's this? :)

Reply:
This was for the requested benchmarking.
It can be disposed of before merging. The benchmarks were added only to answer questions from reviewing the PR.